LUCENE-9385: Add option to skip indexing facet drill down terms

zacharymorn opened the pull request:
# Description / solution

Add option to skip indexing facet drill down terms

# Tests
* added new passing tests

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/lucene/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Lucene maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.

zacharymorn committed:
LUCENE-9385: Add option to skip indexing facet drill down terms

zacharymorn commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -444,12 +478,6 @@ private void processAssocFacetFields(
         FacetsConfig.DimConfig ft = getDimConfig(field.dim);
 
         // Drill down:
-        int start;
```
This `start` variable is not used in code.

mikemccand replied:
Good catch!  It's crazy our static tooling can't ... help us remove such code :)

zacharymorn replied:
Yeah I was surprised it didn't get caught by static analysis as well (my IDE did though).

mikemccand replied:
Maybe open a follow-on issue?  Surely `javac` has some command line flags these days to warn about dead code?  If we turned it on would we be horrified by how much dead we have lying about?

zacharymorn replied:
Sure I've created a jira issue to track this https://issues.apache.org/jira/browse/LUCENE-9856. I see from https://issues.apache.org/jira/browse/LUCENE-9497 `error prone` was already integrated to be used as static analysis tool during compilation, so I guess it may just require some additional configuration.

rmuir commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -65,6 +65,8 @@
   // int/float/bytes in a single indexed field:
   private final Map<String, String> assocDimTypes = new ConcurrentHashMap<>();
 
+  private boolean disableDrillDown = false;
```
Can we rework this to be `enableDrillDown = true`?

zacharymorn replied:
Thanks for the suggestion Robert! I had previously thought about it (prefer positive statement over negative one), but ended up choosing `disableDrillDown` variable / API naming as drill down terms indexing is an existing behavior, so I'm not sure if adding `enableDrillDown` variable / API may add confusion for existing Lucene clients that this indexing needs to be explicitly enabled.  

However, I'm planning to change the API from taking a boolean to taking an enum instead as Mike suggested (`setDrillDownTermsIndexingOption`) , so I guess we should be good there?

mikemccand commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -87,6 +89,8 @@
      * True if drilling down by a whole dimension, to match all documents that had any value for
      * this dimension, is necessary (default is true)
      */
+    // nocommit should this (control for only label / dim name indexing) be deprecated to favor the
```
I think there three use cases we want to enable, which we are now doing with two booleans:

  * Index no drill down terms
  * Index dimension and full path drill down
  * Index only full path drill down

I think the 4th logical option (Index only dimension) is very weird and we don't need to make that easy?

So maybe we could make an `enum` for these three, instead of two booleans?

zacharymorn replied:
Using enum to replace the 2 booleans is a great idea! I'll change it in a bit. 

However, I have been thinking about one question during when I worked on this yesterday, and think I should ask after seeing your comment above: does full path term count as "drill down term" ?  If it does count, and full path term should be skipped for indexing as well when drill down term indexing is disabled, then wouldn't the full path information provided via `FacetField(dimension, path1, path2, path3...)` is lost in the index? If it does not count, and we should always index full path term, then wouldn't `Index no drill down terms` == `Index only full path drill down`?

For clarity, I've listed the enum options with different drill down terms examples below that I have in mind. Please let me know if this is the correct understanding.  

```
  /**
   * Drill down terms indexing option to control whether dimension and sub-path terms should be
   * indexed.
   */
  public enum DrillDownTermsIndexing {
    /**
     * Index only full-path drill down terms. No dimension nor sub-path indexing. 
     * e.g. for FaceField("a", "foo/bar/baz"), we would only index value "a foo bar baz"
     */
    FULL_PATH_ONLY,

    /**
     * Index sub-path and full-path down terms. No dimension indexing. 
     * e.g. for FaceField("a", "foo/bar/baz"), we would only index values "a foo", "a foo bar", and "a foo baz"
     */
    SUB_PATH_AND_FULL_PATH,

    /**
     * Index dimension and full-path drill down terms. No sub-path indexing. 
     * e.g. for FaceField("a", "foo/bar/baz"), we would only index values "a" and "a foo baz" 
     * This is weird usage and currently not available.
     */
    // DIMENSION_AND_FULL_PATH,

    /**
     * Index dimension, sub-path and full-path drill down terms. 
     * e.g. for FaceField("a", "foo/bar/baz"), we would index all values "a", "a foo", "a foo bar", and "a foo baz" 
     * This is the default / existing behavior
     */
    ALL
  }
```

zacharymorn replied:
I've pushed a new commit using enum with above definition for now, and moving the flag from `FacetsConfig` to `DimConfig` so that it's still configurable at the field level. Please let me know if any changes need to be made there.

mikemccand reviewed (commented):
Thank you @zacharymorn!

I left a comment about maybe using an `enum` instead of the two booleans ... I agree it is confusing with two booleans.

gsmiller commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -178,6 +190,25 @@ public synchronized void setRequireDimensionDrillDown(String dimName, boolean v)
 
+  /**
+   * Specify whether drill down functionality as a whole should be disabled. Default to false for
+   * backward compatibility.
```
I wonder if calling out "backward compatibility" here might create confusion. It reads like the "new best thing" is to disable drilling down, which really isn't the case. Seems like just defaulting to enabling drill down is a simple enough explanation (hey, that's what most users probably expect right?).

Also, to echo Robert's comment above, I suggest enableDrillDown instead of focusing on "disabling"

zacharymorn replied:
Thank Greg for the comment! I was coding at night after some beer so might have used the words a bit casually :D , but yes my original meaning there was that the default setting should not break existing Lucene clients upgrading from older version. I'll re-word this java doc entirely as I'll change the API to taking enum instead of boolean. 

For `enableDrillDown` vs. `disableDrillDown`, please see my reply to Robert's comment above.

gsmiller commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -372,7 +403,10 @@ private void processFacetFields(
 
         // Drill down:
```
Hmm, maybe this is just a personal preference, but I'd find it more readable to do:

```suggestion
        // Drill down:
        if (disableDrillDown == false) {
          int start;
          if (ft.requireDimensionDrillDown) {
            start = 1;
          } else {
            start = 2;
          }
          for (int i = start; i <= cp.length; i++) {
            doc.add(new StringField(indexFieldName, pathToString(cp.components, i), Field.Store.NO));
          }
        }
```

zacharymorn replied:
I think this logic may or may not be correct actually, as it depends on whether full path terms is considered as "drill down terms" (see my question to Mike above).

Also, the index `i` in `pathToString(cp.components, i)` above is used as "length". So with the original implementation when `start = cp.length`,  `pathToString(cp.components, start)`  will still be executed and give the full path including every element inside `cp.components`

mikemccand replied:
I think this approach is good, i.e. "disable drill down" would also mean "disable the full path drill down", i.e. it would mean this `dim` adds zero tokens to the inverted index.

zacharymorn replied:
Updated to use the new `DrillDownTermsIndexing.NONE` option to denote the all drill down disabled condition. Also refactored a bit more there to support the `DrillDownTermsIndexing.DIMENSION_AND_FULL_PATH` option that was previously commented out.

zacharymorn committed:
Use enum instead of 2 boolean to control drill down terms indexing

zacharymorn committed:
Fix typo and javadoc

mikemccand commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -65,6 +65,38 @@
+  public enum DrillDownTermsIndexing {
+    /**
+     * Index only full-path drill down terms. No dimension nor sub-path indexing. e.g. for
+     * FacetField("a", "foo/bar/baz"), we would only index value "a foo bar baz".
```
Instead of `"a foo bar baz"` could we consistently use `/`, i.e. change it to `a/foo/bar/baz`?

zacharymorn replied:
Updated. I used `a foo bar baz` before as that's the format computed by `pathToString(cp.components, i)` and fed into the value of `StringField`, but yeah given that is actually internal drill down term indexing format, we should probably not expose that in java doc and favor `a/foo/bar/baz` instead.

mikemccand commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -65,6 +65,38 @@
+    FULL_PATH_ONLY,
+
+    /**
+     * Index sub-path and full-path down terms. No dimension indexing. e.g. for FacetField("a",
```
Insert `drill` before `down terms`?

zacharymorn replied:
Ops. Updated.

mikemccand commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -65,6 +65,38 @@
+
+    /**
+     * Index sub-path and full-path down terms. No dimension indexing. e.g. for FacetField("a",
+     * "foo/bar/baz"), we would only index values "a foo", "a foo bar", and "a foo baz".
```
And put `/` in the examples too, e.g. `a/foo`, `a/foo/bar`, ...

zacharymorn replied:
Updated. I also noticed I had put in some incorrect examples, so I fixed those as well.

mikemccand commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -65,6 +65,38 @@
+     * "foo/bar/baz"), we would index all values "a", "a foo", "a foo bar", and "a foo baz". This is
+     * the default / existing behavior.
+     */
+    ALL
```
Hmm don't we also need a `NONE`?  I.e. disable all drill down indexing entirely (the original goal of the issue I think)?

zacharymorn replied:
Added the `NONE` option.

gsmiller replied:
Love the examples in the doc. Thanks!

mikemccand commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -65,6 +65,38 @@
+     * Index sub-path and full-path down terms. No dimension indexing. e.g. for FacetField("a",
+     * "foo/bar/baz"), we would only index values "a foo", "a foo bar", and "a foo baz".
+     */
+    SUB_PATH_AND_FULL_PATH,
```
Maybe rename to `ALL_PATHS_NO_DIM`?

zacharymorn replied:
Renamed.

mikemccand commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -65,6 +65,38 @@
+     * FacetField("a", "foo/bar/baz"), we would only index values "a" and "a foo baz". This is weird
+     * usage and currently not available.
+     */
+    // DIMENSION_AND_FULL_PATH,
```
Actually, I am now thinking that this case might be helpful?  (It was the "only dim and NOT full path" case that I thought seems pointless) ... this case could e.g. be used to find all docs that have at least one facet label on the dimension.  So, maybe uncomment this and fixup the doc, e.g. add `/` in, remove the weirdness statement?

zacharymorn replied:
Updated. Yeah I wasn't sure before if full-path is considered as drill down terms as well, but the new `NONE` option makes it clear now.

mikemccand reviewed (commented):
Thank you @zacharymorn -- I left some more comments.

zacharymorn committed:
Address comments:
1. Update java doc examples to have consistent format for drill down terms
2. Add NONE optino to also skip full path drill down term
3. Enable DIMENSION_AND_FULL_PATH option

zacharymorn committed:
Add missed assertion

gsmiller commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -371,14 +426,26 @@ private void processFacetFields(
-        for (int i = start; i <= cp.length; i++) {
-          doc.add(new StringField(indexFieldName, pathToString(cp.components, i), Field.Store.NO));
+        if (ft.drillDownTermsIndexing != DrillDownTermsIndexing.NONE) {
+          StringField fullPathField =
```
minor: You could collapse these two lines and skip declaring the local variable fullPathField to be consistent with your other code.

zacharymorn replied:
I introduced `fullPathField` here as I thought it could use some explanation to add clarity, but I'm fine either way so I have inlined it here.

gsmiller commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -371,14 +426,26 @@ private void processFacetFields(
+          if (ft.drillDownTermsIndexing == DrillDownTermsIndexing.DIMENSION_AND_FULL_PATH) {
+            doc.add(
+                new StringField(indexFieldName, pathToString(cp.components, 1), Field.Store.NO));
+          } else if (ft.drillDownTermsIndexing == DrillDownTermsIndexing.ALL_PATHS_NO_DIM) {
+            for (int i = 2; i < cp.length; i++) {
+              doc.add(
+                  new StringField(indexFieldName, pathToString(cp.components, i), Field.Store.NO));
+            }
+          } else if (ft.drillDownTermsIndexing == DrillDownTermsIndexing.ALL) {
+            for (int i = 1; i < cp.length; i++) {
+              doc.add(
+                  new StringField(indexFieldName, pathToString(cp.components, i), Field.Store.NO));
+            }
+          }
```
minor: you might consider the following, which might make it more clear to future readers that the only difference in these two options is the start length:

```suggestion
} else if (ft.drillDownTermsIndexing == DrillDownTermsIndexing.ALL_PATHS_NO_DIM || ft.drillDownTermsIndexing == DrillDownTermsIndexing.ALL) {
            int start;
            if (ft.drillDownTermsIndexing == DrillDownTermsIndexing.ALL) {
              start = 1;
            } else {
              start = 2;
            }
            for (int i = start; i < cp.length; i++) {
              doc.add(new StringField(indexFieldName, pathToString(cp.components, i), Field.Store.NO));
            }
          }
```

zacharymorn replied:
hmmm tbh I actually preferred the existing one for readability as each option is checked once, whereas the suggested one has additional branching to check for one option twice. I see both styles having their pros & cons...

gsmiller replied:
Yeah, that's fair. My thinking was that the loop start being the only difference could be overlooked in a future change, but I don't really have strong opinions here. Thanks for considering a different approach :)

gsmiller commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -408,7 +475,7 @@ private void processSSDVFacetFields(
         doc.add(new StringField(indexFieldName, fullPath, Field.Store.NO));
```
Does this also need to get modified?

zacharymorn replied:
I think this would function like the existing behavior since I made the corresponding changes for setting `requireDimensionDrillDown`:

```
  /**
   * Specify whether drill down on the dimension is necessary.
   *
   * @deprecated Use {@link FacetsConfig#setDrillDownTermsIndexing(String, DrillDownTermsIndexing)}
   *     instead
   */
  @Deprecated
  public synchronized void setRequireDimensionDrillDown(String dimName, boolean v) {
    DimConfig ft = fieldTypes.get(dimName);
    if (ft == null) {
      ft = new DimConfig();
      fieldTypes.put(dimName, ft);
    }

    ft.drillDownTermsIndexing =
        v ? DrillDownTermsIndexing.ALL : DrillDownTermsIndexing.ALL_PATHS_NO_DIM; // drillDownTermsIndexing = true is now drillDownTermsIndexing = DrillDownTermsIndexing.ALL
  }
```

but I guess a relevant question would be, do we want to support different indexing options here as well like in `processFacetFields` ? My understanding is this method is not used for taxonomy faceting, so we may not need to add indexing for drill down terms, but I may be wrong. What do you think @mikemccand ?

gsmiller replied:
Yeah, these are interesting because they can't really honor all the options you've introduced in the new enum right? The enum is pretty centered around the taxonomy use-case, but then these other two sneak in.

These other two (SSDV and AssociationFacet) can honor NONE, FULL_PATH_ONLY and DIMENSION_AND_FULL_PATH, but the other two options are weird because there aren't cases (that I'm aware of anyway) for indexing partial paths. Hmm. Well, at a minimum, I think they should honor those first three options. The current behavior is either DIMENSION_AND_FULL_PATH or FULL_PATH_ONLY, and NONE would be trivial to add. Need a little more thought on the other two. Maybe Mike has some additional ideas :)

mikemccand replied:
Indeed, this is a weird forced-fit: SSDV facets are not hierarchical, or rather, they have only `dim` and `label`.

But I think all of the new `enum` options can be implemented correctly here, as if this were a taxonomy facets field of depth two (`a/b`)?

zacharymorn replied:
Sounds good. I've added support for these options for SSDV as well in the latest commits. Actually during implementation I noticed there are test assertions for drill down for SSDV already, so I guess there's already overlap between the two (do we want to combine the overlap though?).

gsmiller commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -444,12 +511,6 @@ private void processAssocFacetFields(
-          start = 2;
-        }
         for (int i = 1; i <= label.length; i++) {
           doc.add(
```
Does this also need to be modified to honor the config?

zacharymorn replied:
I originally thought it doesn't need to be updated as association facet field doesn't work with hierarchy, and the `int i = 1` in the for loop seems to be done on purpose:

https://github.com/apache/lucene/blob/2678d68be8e1eab75706c3aedf27ca4c77182a22/lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java#L263

But on a second thought / look I think it should use the same configuration now, particularly after `NONE` option is added. I've updated it accordingly. Good catch!

zacharymorn committed:
Address comment to add skip drill down terms option in association faceting

mikemccand commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -388,6 +434,26 @@ private void processFacetFields(
+        for (int i = 1; i < cp.length; i++) {
+          doc.add(new StringField(indexFieldName, pathToString(cp.components, i), Field.Store.NO));
+        }
+      }
```
Maybe insert `} else { throw AssertionError(...)` to catch future accident of adding a new `enum` value and failing to fix this tricky code?

If we switch to `switch` you could add a `default` case and `throw` there.

zacharymorn replied:
Good call for the assertion error for the potentially new and unhandled enum! I've updated it to use `switch` statement.

mikemccand commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -388,6 +434,26 @@ private void processFacetFields(
 
+  private void indexDrillDownTerms(
+      Document doc, String indexFieldName, DimConfig ft, FacetLabel cp) {
+    if (ft.drillDownTermsIndexing != DrillDownTermsIndexing.NONE) {
```
Maybe, keep the outer `if` checking for `not NONE`, but and then add the full-path term here, but then switch to `switch` for the inner three cases?

zacharymorn replied:
Done.

mikemccand commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -388,6 +434,26 @@ private void processFacetFields(
+  private void indexDrillDownTerms(
+      Document doc, String indexFieldName, DimConfig ft, FacetLabel cp) {
+    if (ft.drillDownTermsIndexing != DrillDownTermsIndexing.NONE) {
+      doc.add(
```
Could you add a comment above this e.g. `// index full-path token` or so?

zacharymorn replied:
Added.

mikemccand commented on lucene/facet/src/test/org/apache/lucene/facet/TestDrillDownQuery.java:
```diff
@@ -299,4 +300,233 @@ public void testRequireDimensionDrillDown() throws Exception {
+    IOUtils.close(taxoReader, reader, writer, dir, taxoDir);
+  }
+
+  public void testDrillDownTermsDefaultWithHierarchicalSetting() throws Exception {
```
Thank you @zacharymorn for the added test cases!

mikemccand commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -168,14 +203,34 @@ public synchronized void setIndexFieldName(String dimName, String indexFieldName
+   *     instead
+   */
+  @Deprecated
   public synchronized void setRequireDimensionDrillDown(String dimName, boolean v) {
```
We should have full backwards compatibility with this, right?  In which case we could backport this change to 8.x too!

zacharymorn replied:
I think we do. The implementation change in `setRequireDimensionDrillDown` should behave like the existing one (keeping or skipping indexing for dim only), existing tests passed (in particular `testRequireDimensionDrillDown`) and there are new tests added to verify the default scenarios (e.g. `testDrillDownTermsDefaultWithHierarchicalSetting`).  Is back-porting something only Lucene committers can do, or I can help out as well?

mikemccand replied:
> Is back-porting something only Lucene committers can do, or I can help out as well?

Oh feel free to help!  That would be very welcome :)  It's a bit tricky right now since Lucene 9.x dev is happening in different repository than 8.x.  Not sure if the normal cherry-pick approach "just works"?

zacharymorn replied:
Sounds good! I assume for 8.x in different repo you are referring to this branch in the old repo right https://github.com/apache/lucene-solr/tree/branch_8x ? And I guess back-porting just means I need to initiate a separate PR with this change, but against that branch / repo, and the PR will go through the same review and approval process to be merged right ?

mikemccand replied:
Yes, that is exactly right @zacharymorn!  So the PR would appear in the old (`lucene-solr`) github repository, not here.  Thank you!

mikemccand replied:
Aha, looks like you already opened the backport PR!  https://github.com/apache/lucene-solr/pull/2471 Thanks @zacharymorn!

mikemccand reviewed (commented):
This is looking close; thanks @zacharymorn!

And thank you for opening the spinoff issue https://issues.apache.org/jira/browse/LUCENE-9856 -- it now has a life all its own and a massive PR fixing many "I see dead code!" problems: https://github.com/apache/lucene/pull/34#pullrequestreview-618569001

zacharymorn committed:
Address comment to support drill down terms indexing option for SSDV faceting as well, and use switch statement

zacharymorn committed:
Merge branch 'main' into LUCENE-9385-Skip-indexing-facet-drill-down-terms

zacharymorn committed:
Small in-lining refactoring

zacharymorn commented:
> This is looking close; thanks @zacharymorn!
> 
> And thank you for opening the spinoff issue https://issues.apache.org/jira/browse/LUCENE-9856 -- it now has a life all its own and a massive PR fixing many "I see dead code!" problems: [#34 (review)](https://github.com/apache/lucene/pull/34#pullrequestreview-618569001)

No problem and it's my pleasure. Thank you for suggesting opening that issue as well!

gsmiller commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -388,6 +434,42 @@ private void processFacetFields(
   }
 
+  private void indexDrillDownTerms(
+      Document doc, String indexFieldName, DimConfig ft, FacetLabel cp) {
```
Does anyone know what "ft" and "cp" actually stand for? Is this a case where the class names changed at some point but the variable names stayed the same? I dunno. Either way, is it time to maybe name these something more descriptive in this method?

zacharymorn replied:
Good catch! For some reasons I've been looking at them the whole time without realizing I should change them. I've updated them in the latest commit and also remove some commented out `System.out.println()`. For the original naming though I'm guess `ft` stands for `fieldType`, but not sure what `cp` stands for.

gsmiller replied:
My money's on "cp == category path" :)

zacharymorn replied:
I'm 99% sure that's a correct guess.

gsmiller commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -388,6 +434,42 @@ private void processFacetFields(
+          new StringField(indexFieldName, pathToString(cp.components, cp.length), Field.Store.NO));
+
+      switch (ft.drillDownTermsIndexing) {
+        case NONE:
```
minor: Should we consider not checking the NONE case here since it's already checked by the outer conditional? I'm surprised the IDE doesn't bark about this actually.

zacharymorn replied:
So not checking `NONE` in this switch statement will actually cause this case (and the `FULL_PATH_ONLY` case) to fall through and reach the default branch, where an assertion error would be raised. I've considered not using `switch` due to these additional checks, but overall I still find them to be more readable than the original if-else-if-else branches.

gsmiller replied:
Got it, thanks!

rmuir replied:
@gsmiller @zacharymorn we may also choose to enable some ECJ checks for switch statements. I will do a quick test of this to see how difficult it would be to turn some of these on.

rmuir replied:
OK I looked, I think it would not be difficult. To improve this, we just have to make decisions about how we want switch-on-enums to work in our codebase.

For example, we could enable `missingEnumCaseDespiteDefault` which would mean that every enum value must be explicitly listed in every switch, even if a `default` case exists. As most enums are small, it would not be that annoying and would ensure compile-time safety if new enum values were added later.

This change would only impact switch statements on enums, so it might be worth looking into. We could still, separately, choose to enforce other things about switch statements in general such as requiring `default` case if we choose to do that, and it wouldn't break this check: we could just make it the pattern to throw an AssertionError from there for enums: all the values are known at compile-time, so we enforce no surprises. 
And requiring `default` is good practice for switches on every other data type because all values are not known.

zacharymorn replied:
> OK I looked, I think it would not be difficult. To improve this, we just have to make decisions about how we want switch-on-enums to work in our codebase.
> 
> For example, we could enable `missingEnumCaseDespiteDefault` which would mean that every enum value must be explicitly listed in every switch, even if a `default` case exists. As most enums are small, it would not be that annoying and would ensure compile-time safety if new enum values were added later.
> 
> This change would only impact switch statements on enums, so it might be worth looking into. We could still, separately, choose to enforce other things about switch statements in general such as requiring `default` case if we choose to do that, and it wouldn't break this check: we could just make it the pattern to throw an AssertionError from there for enums: all the values are known at compile-time, so we enforce no surprises.
> And requiring `default` is good practice for switches on every other data type because all values are not known.

Thanks Robert for the suggestion on this! I like both enforcement ideas. I gave `missingEnumCaseDespiteDefault` a try and can see a few need to be added already. Do you mind I create a spin-off issue for that and work on it separately?

rmuir replied:
please, of course it is best to do any change like this with a separate issue/PR!

zacharymorn replied:
Thanks Robert! I've opened a issue available here https://issues.apache.org/jira/browse/LUCENE-9883

mikemccand replied:
Awesome, thanks @zacharymorn and @rmuir!

gsmiller commented on lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:
```diff
@@ -388,6 +434,42 @@ private void processFacetFields(
     }
   }
 
+  private void indexDrillDownTerms(
```
Refactoring this functionality into its own method makes it so much easier to read. Thanks!

zacharymorn committed:
Address comment for variable naming, and remove some commented out code

mikemccand reviewed (approved):
This looks awesome!  Thank you for fixing all the silly `ft` and `cp` names too!

zacharymorn commented:
> This looks awesome! Thank you for fixing all the silly `ft` and `cp` names too!

Thanks for the review and approval!
